doc/users-guide/users-guide-ipsec.adoc

```diff
@@ -0,0 +1,459 @@
+or that each packet has a specified individual SA.
+
+For outbound lookaside operations, a corresponding `odp_ipsec_out_param_t`
+serves a similar role, but here the SA must be specified since the input
+packet(s) are non-IPsec packets. Again the option is to use a single SA for
+all input packets or one per input packet.
+
+For outbound operations, an associated array of `odp_ipsec_out_opt_t` structs
+is also used to control the fragmentation mode to be used as part of the
+outbound processing. Options here are to not fragment, to fragment before
+IPsec processing, after IPsec processing, or to only check whether IP
+fragmentation is needed but not to perform it. For check processing, the `mtu`
+status error bit in the `odp_ipsec_packet_result_t` is set if check processing
+detects that the resulting packet will not fit into the configured MTU. Note
+that the MTU associated with a given SA is set at SA creation and can be
+changed at any time via the `odp_ipsec_sa_mtu_update()` API.
+
+Once an asynchronous lookaside operation has been initiated, the worker thread
+that issued the asynchronous call can handle other events while waiting for
+the operation to complete. Completion of an asynchronous operation is
+indicated by the worker receiving an `ODP_EVENT_PACKET` that has subtype
+`ODP_EVENT_PACKET_IPSEC`. These events can be retrieved directly by polling
+the completion queue associated with the SA, or (more typically) via the ODP
+scheduler. Typical code for such completion processing would look as follows:
+
+[source,c]
+-----
+while (1) {
+	ev = odp_schedule(&queue, ODP_SCHED_WAIT);
+	ev_type = odp_event_types(ev, &ev_subtype);
+
+	switch (ev_type) {
+	case ODP_EVENT_PACKET:
+
+		switch (ev_subtype) {
+		case ODP_EVENT_PACKET_IPSEC:
+			pkt = odp_packet_from_event(ev);
+
+			if (odp_unlikely(odp_ipsec_result(&result, pkt) != 0)) {
+				/* Stale event, discard */
+				odp_event_free(ev);
+				continue;
+			}
+
+			if (odp_unlikely(result.status.all != ODP_IPSEC_OK)) {
+				 if (result.status.error != ODP_IPSEC_OK) {
+					 ...process error result
+					 odp_event_free(ev);
+					 continue;
+				 } else {
+					 ...process packet warnings
+				 }
+			}
+
+			my_context = odp_ipsec_sa_context(result.sa);
+
+			if (result.flag.inline_mode) {
+				...process inline inbound packet
+			} else {
+				...process the async completion event
+			}
+
+			...
+			break;
+
+		case ...
+		}
+		break;
+
+	case ODP_EVENT_IPSEC_STATUS:
+		...process IPsec status event
+		break;
+
+	}
+}
+-----
+
+===== IPsec Inline Processing
+When IPsec is configured to operate in `ODP_IPSEC_OP_MODE_INLINE` mode,
+inbound processing is implicit. The application never sees these packets until
+after IPsec has already decrypted them. As shown in the code sketch above,
+such packets appear as events of subtype `ODP_EVENT_PACKET_IPSEC` and the
+`flag` field in the associated `odp_ipsec_packet_result_t` indicates
+`inline_mode`.
```

Comment:
May it should be mentioned somewhere here for clarity that also asynchronous look-a-side operations are allowed in the inline mode.